### PR TITLE
Improve `Widget::widget()`

### PR DESCRIPTION
**Pull request by vjik**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ✔️
| Fixed issues  | #50

### CODE REVIEW

**Comment by vjik** [comment]
May be do it differently:

1) `widget()` pass constructor arguments (as in Yii2).

2) `configure()` pass full array configuration.

Seems, in most cases need constructor arguments only.

File: README.md [hunk 1]
```diff
@@ -109,9 +109,16 @@ To set a value for the ID, you can pass it in the configuration array to the `wi
     ],
 ]) ?>
 ```
+Also, you can use shortcut `construct()` with constructor arguments only:
```
Comment (arogachev) [inline README.md:112]:
```suggestion
Also, you can use `construct()` shortcut with constructor arguments only:
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,8 +1,10 @@
 # Yii Widget Change Log
 
-## 1.1.1 under development
+## 2.0.0 under development
 
-- no changes in this release.
+- Chg #72: Change `Widget::widget()` result type to `static` (@vjik)
+- New #72: Add shortcut `Widget::construct()` for create widget by array definition that contain constructor
```
Comment (arogachev) [inline CHANGELOG.md:6]:
```suggestion
- New #72: Add  `Widget::construct()` shortcut for creating widget by array definition that contains constructor
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -1,8 +1,10 @@
 # Yii Widget Change Log
 
-## 1.1.1 under development
+## 2.0.0 under development
 
-- no changes in this release.
+- Chg #72: Change `Widget::widget()` result type to `static` (@vjik)
```
Comment (arogachev) [inline CHANGELOG.md:5]:
```suggestion
- Chg #72: Change `Widget::widget()` result type hint to `static` (@vjik)
```

File: README.md [hunk 4]
```diff
@@ -109,9 +109,16 @@ To set a value for the ID, you can pass it in the configuration array to the `wi
...
 
-For a description of the configuration syntax, see the [yiisoft/factory](https://github.com/yiisoft/factory)
-package documentation.
+```php
+<?= MyWidget::construct([
+    'id' => 'value'
```
Comment (arogachev) [inline README.md:116]:
```suggestion
    'id' => 'value',
```

File: src/Widget.php [hunk 5]
```diff
@@ -96,6 +96,25 @@ final public static function widget(array|callable|string $config = []): self
         return WidgetFactory::createWidget($config);
     }
 
+    /**
+     * Shortcut for {@see self::widget()} with array definition that contain constructor arguments only.
```
Comment (arogachev) [inline src/Widget.php:100]:
```suggestion
     * Shortcut for {@see self::widget()} with array definition that contains constructor arguments only.
```

**Comment by samdark** [comment]
👍 Improves syntax for constructor-only parameters case.
👎 Adds another way to do the same thing thus raising congnitive complexity.

File: src/Widget.php [hunk 6]
```diff
@@ -78,7 +77,8 @@ final public static function end(): string
     /**
      * Creates a widget instance.
      *
-     * @param array|callable|string $config The parameters for creating a widget.
+     * @param array $constructorArguments The constructor arguments.
+     * @param array $config The parameters for creating a widget.
```
Comment (samdark) [inline src/Widget.php:81]: I think we need to referene definitions via `@link` here.

File: src/Widget.php [hunk 7]
```diff
@@ -87,32 +87,19 @@ final public static function end(): string
...
+            $config = [
+                '__construct()' => $constructorArguments,
+            ];
+        } elseif ($constructorArguments !== []) {
+            throw new InvalidArgumentException();
         }
 
-        return WidgetFactory::createWidget($config);
-    }
+        $config['class'] = static::class;
```
Comment (samdark) [inline src/Widget.php:102]: It was previously possible to override class via `class` in config passed explicitly.

**Comment by samdark** [comment]
Need a changelog and adjustments in other packages using widgets.

File: src/Widget.php (hunk 7 above)
Comment (vjik) [reply src/Widget.php:102]: Seems, it's bad hack. And now it throws error, because return type of `widget()` changed to `static`.

**Comment by vjik** [comment]
> Need a changelog and adjustments in other packages using widgets.

Changelog exist

File: src/Widget.php (hunk 6 above)
Comment (vjik) [reply src/Widget.php]: done

**Review by samdark — APPROVED** [review]

**Pull request merged**